### PR TITLE
ft(auth.tsx):Allow users to login with email and password[187364806]

### PR DESCRIPTION
**Pull request by Ke-bean**
To be able to use the application and access authenticated routes, a login form should be presented to users and users should be able to login with email and password

This Feature can be tested by clicking on login form and then enter your credentials and click on sign in with email:

Screen Shot 2024-06-27 at 19 31 17

### CODE REVIEW

**Review by tuyishimehope — DISMISSED** [review]
LGTM

File: src/components/usercomponents/LoginForm.tsx [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+import { zodResolver } from '@hookform/resolvers/zod';
+import { setToken, setUser } from '../../redux/features/auth/authSlice';
+import { Link } from 'react-router-dom';
+import { useForm, SubmitHandler } from 'react-hook-form';
+import { jwtDecode } from 'jwt-decode';
+import { CustomJwtPayload } from '../../redux/features/auth/authSlice';
+import { useNavigate, useLocation, useSearchParams } from 'react-router-dom';
+import { useSelector, useDispatch } from 'react-redux';
+import { useLoginUserMutation } from '../../service/authApi';
+const LoginForm = () => {
```
Comment (shebz2023) [inline src/components/usercomponents/LoginForm.tsx]: import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

File: src/components/usercomponents/LoginForm.tsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    setError,
+  ]);
+
+  const onSubmit: SubmitHandler<FormData> = async (userCredentials) => {
+    await loginUser(userCredentials);
+  };
+  return (
+    <>
+      <div className="flex justify-center items-center flex-col w-2/3 max-h-screen">
+        <form className="flex flex-col bg-white-100  px-8 pt-6 pb-6 my-2 w-full max-w-md">
```
Comment (shebz2023) [inline src/components/usercomponents/LoginForm.tsx]: "ToastContainer />"

**Review by shebz2023 — COMMENTED** [review]
Great job !@Ke-bean ,its better if we return the response to the user, 'toast'library may help

**Review by shebz2023 — DISMISSED** [review]
LGTM

**Review by martinemahirwe — DISMISSED** [review]
LGTM

File: src/components/usercomponents/LoginForm.tsx [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            <button className="bg-black hover:bg-gray-500  font-bold py-1 px-4 rounded focus:outline-none focus:shadow-outline bg-black-500  text-white mb-2 w-full rounded-lg my-2 tracking-wider font-light py-2.5 text-md">
+              Go Back
+            </button>
+            {isSuccess && (
+              <p className="text-green-600 text-xs font-normal">
+                Login successful!
+              </p>
+            )}
+            <h1 className="my-5  text-gray-500 ">or continue with</h1>
+            <button className="bg-black hover:bg-gray-500  font-bold py-1 px-4 rounded focus:outline-none focus:shadow-outline hover:bg-black hover:text-white w-full bg-gray-100 text-black flex flex-row text-gray-600">
```
Comment (SergeInGithub) [inline src/components/usercomponents/LoginForm.tsx]: These buttons almost have the same classes. Can't you make this a component or use the already existing button component and extra classes needed through the className prop?

File: src/components/rootcomponents/Input.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+export const Input: React.FC<InputProps> = ({
+  id,
+  type,
+  placeholder,
+  className,
+  ...rest
+}) => {
+  return (
+    <div className={`mb-4 ${className}`}>
```
Comment (SergeInGithub) [inline src/components/rootcomponents/Input.tsx]: I think we might need this className more on the input than here. Or you can add another one

**Review by SergeInGithub — CHANGES_REQUESTED** [review]
Look into the reviews below @Ke-bean

**Review by tuyishimehope — DISMISSED** [review]
LGTM

**Review by tuyishimehope — DISMISSED** [review]
LGTM

**Review by shebz2023 — DISMISSED** [review]
LGTM

**Review by shebz2023 — DISMISSED** [review]

**Review by martinemahirwe — DISMISSED** [review]
LGTM

**Review by tuyishimehope — DISMISSED** [review]
LGTM

**Review by tuyishimehope — DISMISSED** [review]
LGTM

**Review by shebz2023 — DISMISSED** [review]
LGTM

File: src/components/Input.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+const Input = ({ type, name, title, placeholder, id }: Attributes) => {
+  return (
+    <input
+      type={type}
+      name={name}
+      title={title}
+      id={id}
+      placeholder={placeholder}
+      className="rounded-[8px] w-[60%] border border-[#E0E0E0] py-2 px-4 box-border text-[#828282] shadow-sm focus:outline-none font-medium text-[16px]"
```
Comment (SergeInGithub) [inline src/components/Input.tsx]: I believe we had a className prop here. if not add it

File: src/components/rootcomponents/Header.tsx [hunk 6]
```diff
@@ -10,12 +12,20 @@ export const Header: React.FC = () => {
...
-          className="bg-white text-blue-500 px-4 py-2 rounded hover:bg-gray-200"
-        >
-          Login
-        </Link>
+        {isAuthenticated ? (
+          <button
+            className="bg-white text-blue-500 px-4 py-2 rounded hover:bg-gray-200"
+          >
+            Logout
+          </button>
```
Comment (SergeInGithub) [inline src/components/rootcomponents/Header.tsx]: We have a button component

File: src/components/rootcomponents/LoginInput.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
+      />
+      {errors[id] && typeof errors[id].message === 'string' && (
+        <p className="text-red-500 text-md">{errors[id].message}</p>
+      )}
+    </div>
+  );
+};
+
+export default Input;
```
Comment (SergeInGithub) [inline src/components/rootcomponents/LoginInput.tsx]: Can't we have one input which can be reusable everywhere? Why make another component here when we already had it?

File: src/hooks/useForgotPassword.ts [hunk 8]
```diff
@@ -1,48 +1,48 @@
...
+     const response = await fetch(
+       `${API_BASE_URL}/api/v1/auth/forgot-password`,
+       {
+         method: 'POST',
+         headers: {
+           'Content-Type': 'application/json',
+         },
+         body: JSON.stringify({ email }),
+       },
+     );
```
Comment (SergeInGithub) [inline src/hooks/useForgotPassword.ts]: This is not how we're communicating with the backend. we have redux for this

**Review by SergeInGithub — CHANGES_REQUESTED** [review]
Nice work so far. Some reviews below @Ke-bean

**Review by martinemahirwe — DISMISSED** [review]
good work but you have to work on TTL suggestions about best practices as well @Ke-bean

**Review by tuyishimehope — DISMISSED** [review]
LGTM

**Review by shebz2023 — APPROVED** [review]
LGTM

**Review by tuyishimehope — APPROVED** [review]
LGTM

**Review by martinemahirwe — APPROVED** [review]
LGTM

**Review by SergeInGithub — APPROVED** [review]
LGTM

**Pull request merged**